docs/source/app-dev/command-deduplication.rst

```diff
@@ -0,0 +1,339 @@
+  
+#. The :ref:`application ID <com.daml.ledger.api.v1.Commands.application_id>` identifies the application that submits the command.
+
+#. The :ref:`command ID <com.daml.ledger.api.v1.Commands.command_id>` is chosen by the application to identify the intended ledger change.
+
+#. The :ref:`deduplication period <com.daml.ledger.api.v1.Commands.deduplication_period>` specifies the period for which no earlier submissions with the same change ID should have been accepted, as witnessed by a completion event on the :ref:`command completion service <command-completion-service>`.
+   Otherwise, the current submission shall be rejected.
+   The period is specified either as a :ref:`deduplication duration <com.daml.ledger.api.v1.Commands.deduplication_duration>` or as a :ref:`deduplication offset <com.daml.ledger.api.v1.Commands.deduplication_offset>`.
+
+#. The :ref:`submission ID <com.daml.ledger.api.v1.Commands.submission_id>` is returned to the submitting application and allows it to correlate specific submissions to specific completions.
+   An application should never reuse a submission ID.
+
+The ledger may arbitrarily extend the deduplication period specified in the submission, even beyond the maximum deduplication time specified in the :ref:`ledger configuration <ledger-configuration-service>`.
+The deduplication period chosen by the ledger is the *effective deduplication period*.
+The ledger may also convert a requested deduplication duration into an effective deduplication offset or vice versa.
+The effective deduplication period is reported in the the command completion event in the :ref:`deduplication duration <com.daml.ledger.api.v1.Completion.deduplication_duration>` or :ref:`deduplication offset <com.daml.ledger.api.v1.Completion.deduplication_offset>` fields.
+
+A command submission is considered a **duplicate submission** if at least one of the following holds:
+
+- The submitting participant's completion service contains a successful completion event for the same :ref:`change ID <change-id>` within the *effective* deduplication period.
+
+- The participant or Daml ledger are aware of another command submission in-flight with the same :ref:`change ID <change-id>` when they performs command deduplication.
+
+Command submissions via the :ref:`command service <command-service>` indicate the command deduplication outcome as a synchronous gRPC response (except when the gRPC deadline was exceeded before).
+Submissions via the :ref:`command submission service <command-submission-service>` can indicate the outcome synchronously or asynchronously in the event on the :ref:`command completion service <command-completion-service>`.
+In particular, the submission may be a duplicate even if the command submission service acknowledges the submission with the gRPC status code ``OK``.
+Command deduplication generates the following outcomes of a command submission:
```

Comment:
It's unclear to me whether the following list only applies to completion events, or if the same errors can also be returned synchronously by the submission service.

Reply:
The sentences above this one say that outcomes are reported either synchronously or asynchronously. I've moved the sentence to the beginning of the paragraph. Does that help with the unclarity?

Reply:
Thanks for the suggested structure. I've added it, with a note saying that the synchronous completion of the command service may also appear on the completion stream (that's apparently something that some readers of our docs were surprised to find out).